### PR TITLE
Immutable params

### PR DESCRIPTION
**Pull request by lawhead**
# Overview

In the current workflow, any edits to the configuration for a session overwrite the default parameters. This makes it problematic to upgrade BciPy since this file often gets updated with new configuration options. This same file is used for both defining which parameters are available and their default values, as well as the configuration used for the current session.

This PR separates out these two concepts so that the default parameters.json file never gets overridden when configuration a session. The workflow has been modified so that users in the RSVPKeyboard GUI interface can select an existing parameters file. If they attempt to edit the configuration without first selecting experiment parameters, a new parameters file is created by copying the default params.

## Ticket

https://www.pivotaltracker.com/story/show/175193758

## Contributions

- Modified RSVPKeyboard GUI to change the workflow for selecting and editing a parameters file. The configured file is passed to both the params_form for editing and bci_main for starting a session.
- Added a command line argument to bci_main to accept a parameters file location.
- Added functionality for copying a parameters file.
- Created a new Parameters class which extends dict. Usage throughout the codebase for accessing configuration is unchanged, but it retains metadata and can be saved as a JSON file in our expected serialization format. It also performs validation of configuration parameters.

## Test

- Run the unit tests to ensure parameter functionality is unchanged.
- In the RSVPKeyboard GUI:
   - Attempt to configure parameters when none has been selected. A new parameters file with the current timestamp should be created. After starting a calibration session, the parameters file in the data folder should match the new params file. The parameter_location property should point to the new file.
  - Load an existing parameters file.

### CODE REVIEW

File: bcipy/helpers/load.py [hunk 1]
```diff
@@ -1,56 +1,49 @@
...
+from shutil import copyfile
+from pathlib import Path
 import pickle
+from bcipy.helpers.parameters import Parameters, DEFAULT_PARAMETERS_PATH
 
 from tkinter.filedialog import askopenfilename, askdirectory
 
 log = logging.getLogger(__name__)
 
+DEFAULT_PARAMETERS_PATH = 'bcipy/parameters/parameters.json'
```
Comment (tab-cmd) [inline bcipy/helpers/load.py]: It looks like we should pull this from from `bcipy.helpers.parameters`?

File: bcipy/helpers/load.py [hunk 2]
```diff
@@ -1,56 +1,49 @@
...
-    for key, value in parameters.items():
-        new_parameters[key] = cast_value(value)
-
-    return new_parameters
-
-
-def cast_value(value):
-    """Cast Value.
+def copy_parameters(path: str = DEFAULT_PARAMETERS_PATH,
+                    destination: str = 'bcipy/parameters/') -> str:
```
Comment (tab-cmd) [inline bcipy/helpers/load.py]: Let's make this a constant or derived from `DEFAULT_PARAMETERS_PATH `

File: bcipy/helpers/tests/test_save.py [hunk 3]
```diff
@@ -12,7 +13,7 @@ class TestSave(unittest.TestCase):
     def setUp(self):
         # set up the needed paths and initial data save structure
 
-        self.data_save_path = 'data/'
+        self.data_save_path = tempfile.mkdtemp()
```
Comment (tab-cmd) [inline bcipy/helpers/tests/test_save.py:16]: 👍

File: bcipy/helpers/parameters.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+from typing import Tuple
+from codecs import open as codecsopen
+from collections import abc
+from json import dump, load
+from pathlib import Path
+
+DEFAULT_PARAMETERS_PATH = 'bcipy/parameters/parameters.json'
+
+
+class Parameters(dict):
```
Comment (tab-cmd) [inline bcipy/helpers/parameters.py:11]: This makes me really happy. I like the parameters being a class that can handle itself.

File: bcipy/helpers/load.py [hunk 5]
```diff
@@ -1,56 +1,49 @@
...
-            new_value = True if actual_value == 'true' else False
-        elif actual_type == 'str' or 'path' in actual_type:
-            new_value = str(actual_value)
-        else:
-            raise ValueError('Unrecognized value type')
-
-    except Exception:
-        raise ValueError(f'Could not cast {actual_value} to {actual_type}')
+    now = datetime.now()
+    month = str(now.month).rjust(2, "0")
```
Comment (tab-cmd) [inline bcipy/helpers/load.py]: I think it's cleaner to use `strftime` for filenames with datetimes. I use it in `helpers/save`

File: bcipy/gui/params_form.py [hunk 6]
```diff
@@ -360,16 +370,24 @@ def OnChildFocus(self, event):
...
-    fa = MainPanel(frame, title=title, json_file=json_file)
+    fa = MainPanel(frame, json_file=json_file, title=title)
     frame.Show()
     app.MainLoop()
 
 
 if __name__ == '__main__':
-    main()
+    import argparse
+    from bcipy.helpers.load import DEFAULT_PARAMETERS_PATH
```
Comment (tab-cmd) [inline bcipy/gui/params_form.py]: I'd be okay loading from `bcipy.helpers.parameters`

File: bcipy/gui/params_form.py [hunk 7]
```diff
@@ -311,17 +321,17 @@ def __init__(self, parent, title='BCI Parameters',
...
                                             size=14))
+        loading_box.AddSpacer(7)
         self.loaded_from = static_text_control(
             self,
-            label=f'Loaded from: {json_file}',
+            label='Set values from another file',
             size=14)
         loading_box.Add(self.loaded_from)
-        loading_box.AddSpacer(10)
+        # loading_box.AddSpacer(5)
```
Comment (tab-cmd) [inline bcipy/gui/params_form.py]: feel free to remove if commented out code isn't needed

File: bcipy/gui/mode/RSVPKeyboard.py [hunk 8]
```diff
@@ -45,8 +73,9 @@ def launch_bci_main(self, event: wx.Event) -> None:
             username = self.comboboxes[0].GetValue().replace(" ", "_")
             experiment_type = event.GetEventObject().GetId()
             mode = 'RSVP'
-            cmd = 'python bci_main.py -m {} -t {} -u {}'.format(
-                mode, experiment_type, username)
+            # TODO: add -p flag and use configured parameter_location
```
Comment (tab-cmd) [inline bcipy/gui/mode/RSVPKeyboard.py]: looks like you're doing that?

File: bcipy/gui/mode/RSVPKeyboard.py [hunk 9]
```diff
@@ -1,27 +1,33 @@
...
 import itertools
 import os
 import subprocess
 
 import wx
 
 from bcipy.gui.gui_main import BCIGui
-from bcipy.helpers.load import load_json_parameters
+from bcipy.helpers.load import load_json_parameters, copy_parameters
+from bcipy.helpers.load import DEFAULT_PARAMETERS_PATH
```
Comment (tab-cmd) [inline bcipy/gui/mode/RSVPKeyboard.py]: I'd take this from `helpers/parameters` or keep it on the same line

**Review by tab-cmd — APPROVED** [review]
A few comments to address. Let's get time with one of the clinical team members (Betts, Deirdre, or Dan) to review the GUI changes.

File: bcipy/helpers/load.py (hunk 5 above)
Comment (lawhead) [reply bcipy/helpers/load.py]: Thanks, I'll look at that.

File: bcipy/helpers/load.py (hunk 1 above)
Comment (lawhead) [reply bcipy/helpers/load.py]: Yes, good point. I'll clean that up.

File: bcipy/gui/mode/RSVPKeyboard.py (hunk 8 above)
Comment (lawhead) [reply bcipy/gui/mode/RSVPKeyboard.py]: Oh, yep. I'll remove this.

File: bcipy/gui/params_form.py (hunk 7 above)
Comment (lawhead) [reply bcipy/gui/params_form.py]: Thanks, good catch.

**Pull request merged**